### PR TITLE
JAMES-2733 Deleted MailQueue elements should be dequeued

### PR DESCRIPTION
**Pull request by chibenwa**
Let's accuse flatMap VS concatMap... :trollface:

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[ERROR]   RabbitMQEventBusTest$LifeCycleTest$SingleEventBus$DispatchingWhenNetWorkIssue.dispatchShouldWorkAfterNetworkIssuesForOldRegistration:333 » ConditionTimeout
```

test this please

**Comment by Arsnael** [comment]
```
[d8729b524d32ddde2d1e7e7a0c8e0c8141d48745] [ERROR]   ManageableMailQueueContract.dequeueShouldDecreaseQueueSize:82 » AlreadyClosed ...
[d8729b524d32ddde2d1e7e7a0c8e0c8141d48745] [ERROR]   ManageableMailQueueContract.noAckShouldNotDecreaseSize:93 » AlreadyClosed chan...
```
test this please

**Comment by chibenwa** [comment]
```
[d8729b524d32ddde2d1e7e7a0c8e0c8141d48745] [ERROR]   ManageableMailQueueContract.dequeueShouldDecreaseQueueSize:82 » AlreadyClosed ...
[d8729b524d32ddde2d1e7e7a0c8e0c8141d48745] [ERROR]   ManageableMailQueueContract.noAckShouldNotDecreaseSize:93 » AlreadyClosed chan...
```

**Comment by chibenwa** [comment]
Clearly, the fixes I did do address the symptoms but not the root cause and I predict bad things to happens as long as the root cause is not fixed...

Here is an email I just sent to server-dev :

```
While working on a bug solving of the RabbitMQ MailQueue here https://github.com/linagora/james-project/pull/2445 (to recall it: deleted emails are still dequeues, which was not tested by the MailQueueContract), I realized one of the invariants we enforced on MailQueue does not hold.

The invariant is: "A single email is enqueued a single time in a given queue".

That invariant is broken for instance when:
 - reprocessing an email: the name is not altered
 - Recipient RewriteTable rewrites to a distant server
 - RemoteDelivery bouncing configured with an SMTP gateway

RabbitMQ mail queue does need to leverage a Cassandra projection in order to offer MailQueue manageable capabilities (clear, delete, browse, size amongst others). We naturally chose the mail name, used everywhere to identify an email in order to build this projection.

This results in a given mail being enqueued a single time in a single queue: on the second enqueue, the mail will be referenced as already deleted from the queue and will be discarded.

In https://github.com/linagora/james-project/pull/2446 I scratched the surface by enforcing the aforementioned invariants where the right approach is to make the RabbitMQ mail queue correctly handle multiple emails enqueue with the same name. Which might involve table schema changes.

Any thoughts on this topic?
```

File: server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java [hunk 1]
```diff
@@ -80,21 +81,33 @@ public void done(boolean success) {
             .filter(getResponse -> getResponse.getBody() != null);
     }
 
-    Flux<MailQueue.MailQueueItem> deQueue() {
-        return flux.flatMap(this::loadItem);
+    Flux<? extends MailQueue.MailQueueItem> deQueue() {
+        return flux.flatMap(this::loadItem)
+            .concatMap(this::filterIfDeleted);
```
Comment (blackheaven) [inline server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java]: I do not see the difference with the last attempt which broke up some tests.

Comment (chibenwa) [reply server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java]: Yes this fix can [should] be discarded

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/2454

**Pull request closed without merging**